pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/MissingOverrideRule.java

```diff
@@ -0,0 +1,363 @@
+    }
+
+
+    private Set<Method> overriddenMethodsRec(Class<?> exploredType, boolean skip, Set<Method> candidates, Set<Method> result, Set<Class<?>> alreadyExplored) {
+
+        if (candidates.isEmpty() || alreadyExplored.contains(exploredType)) {
+            return result;
+        }
+
+        alreadyExplored.add(exploredType);
+
+        if (!skip) {
+            Set<Method> toRemove = new HashSet<>();
+            for (Method dm : exploredType.getDeclaredMethods()) {
+                for (Method cand : candidates) {
+                    if (cand.getName().equals(dm.getName()) && Arrays.equals(cand.getParameterTypes(), dm.getParameterTypes())) {
+                        // cand is overriden
+                        result.add(cand);
+                        toRemove.add(cand);
+                        // Several methods are eligible, because of return type covariance
+                        // We could do away with adding only the first one to the result,
+                        // but then the other would stay in the candidates and we'd explore
+                        // the rest of the tree unnecessarily
+                    }
+                }
+                candidates.removeAll(toRemove); // no need to look for it elsewhere
+            }
+        }
+
+        if (candidates.isEmpty()) {
+            return result;
+        }
+
+        Class<?> superClass = exploredType.getSuperclass();
+        if (superClass != null) {
+            overriddenMethodsRec(superClass, false, candidates, result, alreadyExplored);
+        }
+
+        for (Class<?> iface : exploredType.getInterfaces()) {
+            overriddenMethodsRec(iface, false, candidates, result, alreadyExplored);
+        }
+
+        if (exploredType.isInterface() && exploredType.getInterfaces().length == 0) {
+            // implicit object member declarations
+            if (!alreadyExplored.contains(Object.class)) {
+                overriddenMethodsRec(Object.class, false, candidates, result, alreadyExplored);
+            }
+        }
+
+        return result;
+    }
+
+
+    @Override
+    public Object visit(ASTMethodDeclaration node, Object data) {
+        if (currentLookup.peek() == null) {
+            return super.visit(node, data);
+        }
+
+        for (ASTAnnotation annot : node.getDeclaredAnnotations()) {
+            if (Override.class.equals(annot.getType())) {
+                // we assume the compiler has already checked it, so it's correct
+                return super.visit(node, data);
+            }
+        }
+
+        try {
+            boolean overridden = currentLookup.peek().isOverridden(node.getName(), node.getFormalParameters());
+            if (overridden) {
+
+                addViolation(data, node, new Object[]{node.getQualifiedName().getOperation()});
+            }
+        } catch (NoSuchMethodException e) {
```

Comment:
Since I started the dogfood task, I know, that this EmptyCatchBlock will be flagged. If you are sure, that the exception can be ignored, then use `ignored` or `expected` as the exception variable name.

Reply:
That's a good question, see also #897 . For now, you can log it just via java.util.logging (probably as warning) - and it will probably disappear unnoticed...

If however this exception means, that the further rule execution doesn't make sense, then I would (re)throw it - that way, the exception appears as a processing error in the report.

Reply:
OK thanks, I'll add a logging statement

Atm I know the failure can occur on enum constants declaring a body. We should probably rethrow it and report it as an error, but only once that is fixed in typeres, else these known failures could hide some unknown ones